types/wlr_seat.c

```diff
@@ -372,3 +394,145 @@ void wlr_seat_pointer_send_axis(struct wlr_seat *wlr_seat, uint32_t time,
+		wl_keyboard_send_enter(handle->keyboard, serial,
+			surface->resource, &keys);
+
+		// TODO: send modifiers
+	}
+
+	// reinitialize the focus destroy events
+	wl_list_remove(
+		&wlr_seat->keyboard_state.focus_surface_destroy_listener.link);
+	wl_list_init(&wlr_seat->keyboard_state.focus_surface_destroy_listener.link);
+	wl_list_remove(
+		&wlr_seat->keyboard_state.focus_resource_destroy_listener.link);
+	wl_list_init(
+		&wlr_seat->keyboard_state.focus_resource_destroy_listener.link);
+	if (surface) {
+		wl_signal_add(&surface->signals.destroy,
+			&wlr_seat->keyboard_state.focus_surface_destroy_listener);
+		wl_resource_add_destroy_listener(surface->resource,
+			&wlr_seat->keyboard_state.focus_resource_destroy_listener);
+		wlr_seat->keyboard_state.focus_resource_destroy_listener.notify =
+			handle_keyboard_focus_resource_destroyed;
+		wlr_seat->keyboard_state.focus_surface_destroy_listener.notify =
+			handle_keyboard_focus_surface_destroyed;
+	}
+
+	wlr_seat->keyboard_state.focused_handle = handle;
+	wlr_seat->keyboard_state.focused_surface = surface;
+}
+
+void wlr_seat_keyboard_clear_focus(struct wlr_seat *wlr_seat) {
+	struct wl_array keys;
+	wl_array_init(&keys);
+	wlr_seat_keyboard_enter(wlr_seat, NULL, keys);
+}
+
+static bool wlr_seat_keyboard_has_focus_resource(struct wlr_seat *wlr_seat) {
+	return wlr_seat->keyboard_state.focused_handle &&
+		wlr_seat->keyboard_state.focused_handle->keyboard;
+}
+
+uint32_t wlr_seat_keyboard_send_key(struct wlr_seat *wlr_seat, uint32_t time,
+		uint32_t key, uint32_t state) {
+	if (!wlr_seat_keyboard_has_focus_resource(wlr_seat)) {
+		return 0;
+	}
+
+	uint32_t serial = wl_display_next_serial(wlr_seat->display);
+	wl_keyboard_send_key(wlr_seat->keyboard_state.focused_handle->keyboard,
+		serial, time, key, state);
+
+	return serial;
+}
+
+void wlr_seat_keyboard_send_modifiers(struct wlr_seat *wlr_seat,
+		uint32_t mods_depressed, uint32_t mods_latched, uint32_t mods_locked,
+		uint32_t group) {
+	uint32_t serial = 0;
+	struct wl_resource *keyboard;
+
+	if (wlr_seat_keyboard_has_focus_resource(wlr_seat)) {
+		serial = wl_display_next_serial(wlr_seat->display);
+		keyboard = wlr_seat->keyboard_state.focused_handle->keyboard;
+		wl_keyboard_send_modifiers(keyboard, serial, mods_depressed,
+			mods_latched, mods_locked, group);
+	}
+
+	if (wlr_seat_pointer_has_focus_resource(wlr_seat) &&
+			wlr_seat->pointer_state.focused_handle->keyboard &&
+			wlr_seat->pointer_state.focused_handle !=
+				wlr_seat->keyboard_state.focused_handle) {
+		if (serial == 0) {
+			serial = wl_display_next_serial(wlr_seat->display);
+		}
+		keyboard = wlr_seat->pointer_state.focused_handle->keyboard;
+
+		wl_keyboard_send_modifiers(keyboard, serial, mods_depressed,
+			mods_latched, mods_locked, group);
+	}
+}
+
+void wlr_seat_keyboard_set_keymap(struct wlr_seat *wlr_seat, int keymap_fd,
+		size_t keymap_size) {
+	// TODO: we probably should wait to send the keymap if keys are pressed
```

Comment:
This raises an interesting problem with pressing keys on keyboard A, then pressing keys on keyboard B before releasing them.